
Disable PDS by default to avoid surprises.

Signed-off-by: Holger Hoffstätte <holger@applied-asynchrony.com>

diff -rup a/init/Kconfig b/init/Kconfig
--- a/init/Kconfig	2018-08-13 03:02:18.333675395 +0200
+++ b/init/Kconfig	2018-08-13 03:03:05.030747968 +0200
@@ -57,7 +57,7 @@ config SCHED_PDS
           TEST so these options are disabled.
 
           Say Y here.
-	default y
+	default n
 
 
 config BROKEN
