
https://gitlab.com/alfredchen/linux-prjc/-/issues/52

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 07f58156cbf6..0808141470a8 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -2982,7 +2982,7 @@ int sched_fork(unsigned long clone_flags, struct task_struct *p)
	return 0;
 }
 
-void sched_post_fork(struct task_struct *p, struct kernel_clone_args *kargs)
+void sched_cgroup_fork(struct task_struct *p, struct kernel_clone_args *kargs)
 {
	unsigned long flags;
	struct rq *rq;
@@ -3025,6 +3025,13 @@ void sched_post_fork(struct task_struct *p, struct kernel_clone_args *kargs)
	raw_spin_unlock_irqrestore(&p->pi_lock, flags);
 }
 
+void sched_post_fork(struct task_struct *p)
+{
+#ifdef CONFIG_UCLAMP_TASK
+	uclamp_post_fork(p);
+#endif
+}
+
 #ifdef CONFIG_SCHEDSTATS
 
 DEFINE_STATIC_KEY_FALSE(sched_schedstats);
