From e3603326458d279336d6770d9ba6de2cd476bdba Mon Sep 17 00:00:00 2001
From: Con Kolivas <kernel@kolivas.org>
Date: Wed, 19 Oct 2016 00:28:24 +1100
Subject: [PATCH] Remove unused function

---
 kernel/sched/MuQSS.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/kernel/sched/MuQSS.c b/kernel/sched/MuQSS.c
index a14225c..18a3b16 100644
--- a/kernel/sched/MuQSS.c
+++ b/kernel/sched/MuQSS.c
@@ -822,11 +822,6 @@ static inline bool deadline_before(u64 deadline, u64 time)
 	return (deadline < time);
 }
 
-static inline bool deadline_after(u64 deadline, u64 time)
-{
-	return (deadline > time);
-}
-
 /*
  * Deadline is "now" in niffies + (offset by priority). Setting the deadline
  * is the key to everything. It distributes cpu fairly amongst tasks of the
