reverted:
--- b/net/ipv4/tcp_output.c
+++ a/net/ipv4/tcp_output.c
@@ -925,6 +925,8 @@
 	if (clone_it) {
 		TCP_SKB_CB(skb)->tx.in_flight = TCP_SKB_CB(skb)->end_seq
 			- tp->snd_una;
+		tcp_rate_skb_sent(sk, skb);
+
 		oskb = skb;
 		if (unlikely(skb_cloned(skb)))
 			skb = pskb_copy(skb, gfp_mask);
@@ -1039,10 +1041,9 @@
 		tcp_enter_cwr(sk);
 		err = net_xmit_eval(err);
 	}
+	if (!err && oskb)
-	if (!err && oskb) {
 		skb_mstamp_get(&oskb->skb_mstamp);
+
-		tcp_rate_skb_sent(sk, oskb);
-	}
 	return err;
 }
 
