diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index e3966fe..2e9843b 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -3612,7 +3612,9 @@ rerun_prev_unlocked:
 
 static inline void sched_submit_work(struct task_struct *tsk)
 {
-	if (!tsk->state || tsk_is_pi_blocked(tsk))
+	if (!tsk->state || tsk_is_pi_blocked(tsk)
+		|| (preempt_count() & PREEMPT_ACTIVE)
+		|| signal_pending_state(tsk->state, tsk))
 		return;
 	/*
 	 * If we are going to sleep and we have plugged IO queued,
