From 07c05587a0ce671b9da0e03fa27211d605b483af Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 18 Aug 2014 23:56:40 +0800
Subject: [PATCH 19/22] bfs: Make grq cacheline aligned.

---
 kernel/sched/bfs.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index af997f6..0ae76ef 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -228,7 +228,11 @@ static struct root_domain def_root_domain;
 #endif /* CONFIG_SMP */
 
 /* There can be only one */
-static struct global_rq grq;
+#ifdef CONFIG_SMP
+static struct global_rq grq ____cacheline_aligned_in_smp;
+#else
+static struct global_rq grq ____cacheline_aligned;
+#endif
 
 static DEFINE_MUTEX(sched_hotcpu_mutex);
 
-- 
2.4.6

