From 0227acbaa4d4f292a6d60b4e7c56542a13b0c79b Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 21 Aug 2014 22:11:31 +0800
Subject: [PATCH 18/22] bfs: Refactor set_rq_task and inline reset_rq_task.

---
 kernel/sched/bfs.c | 7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index 99b6bd8..af997f6 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -3310,13 +3310,10 @@ static inline void set_rq_task(struct rq *rq, struct task_struct *p)
 	rq->rq_mm = p->mm;
 	rq->rq_smt_bias = p->smt_bias;
 #endif
-	if (p != rq->idle)
-		rq->rq_running = true;
-	else
-		rq->rq_running = false;
+	rq->rq_running = (p != rq->idle);
 }
 
-static void reset_rq_task(struct rq *rq, struct task_struct *p)
+static inline void reset_rq_task(struct rq *rq, struct task_struct *p)
 {
 	rq->rq_policy = p->policy;
 	rq->rq_prio = p->prio;
-- 
2.4.6

