From 6e1526e5dace1e9ea31df26adaee3011baccc004 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 25 Jul 2014 14:27:30 +0800
Subject: [PATCH 08/22] bfs: [Fix] Clean up wake_entry in task_struct

---
 include/linux/sched.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index 5083de9..ef8e51c 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1294,8 +1294,10 @@ struct task_struct {
 	unsigned int flags;	/* per process flags, defined below */
 	unsigned int ptrace;
 
-#if defined(CONFIG_SMP) || defined(CONFIG_SCHED_BFS)
+#if defined(CONFIG_SMP) && !defined(CONFIG_SCHED_BFS)
 	struct llist_node wake_entry;
+#endif
+#if defined(CONFIG_SMP) || defined(CONFIG_SCHED_BFS)
 	int on_cpu;
 #endif
 #ifdef CONFIG_SMP
-- 
2.4.6

