From 3d3bbe7a3d4f6e7f7d684973a9179251ae81e616 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 18 Apr 2015 22:52:49 +0800
Subject: [PATCH 02/22] bfs: [Sync] __schedule() and io_schedule_timeout()

v2
Sync with v4.1
---
 kernel/sched/bfs.c | 21 ++++-----------------
 1 file changed, 4 insertions(+), 17 deletions(-)

diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index 46c8f58e..3244afd 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -3363,7 +3363,6 @@ static void __sched __schedule(void)
 	struct rq *rq;
 	int cpu;
 
-need_resched:
 	preempt_disable();
 	cpu = smp_processor_id();
 	rq = cpu_rq(cpu);
@@ -3411,17 +3410,6 @@ need_resched:
 		switch_count = &prev->nvcsw;
 	}
 
-	/*
-	 * If we are going to sleep and we have plugged IO queued, make
-	 * sure to submit it to avoid deadlocks. This usually clears before
-	 * grabbing the lock but still may rarely happen here. */
-	if (unlikely(deactivate && blk_needs_flush_plug(prev))) {
-		grq_unlock_irq();
-		preempt_enable_no_resched();
-		blk_schedule_flush_plug(prev);
-		goto need_resched;
-	}
-
 	update_clocks(rq);
 	update_cpu_clock_switch(rq, prev);
 	if (rq->clock - rq->last_tick > HALF_JIFFY_NS)
@@ -4823,19 +4811,18 @@ EXPORT_SYMBOL_GPL(yield_to);
 
 long __sched io_schedule_timeout(long timeout)
 {
-	struct task_struct *curr = current;
-	int old_iowait = curr->in_iowait;
+	int old_iowait = current->in_iowait;
 	struct rq *rq;
 	long ret;
 
-	curr->in_iowait = 1;
-	blk_schedule_flush_plug(curr);
+	current->in_iowait = 1;
+	blk_schedule_flush_plug(current);
 
 	delayacct_blkio_start();
 	rq = raw_rq();
 	atomic_inc(&rq->nr_iowait);
 	ret = schedule_timeout(timeout);
-	curr->in_iowait = old_iowait;
+	current->in_iowait = old_iowait;
 	atomic_dec(&rq->nr_iowait);
 	delayacct_blkio_end();
 
-- 
2.4.6

