From 92099d80630ac5742447b5c02ec2399bd7c248fe Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 15 Oct 2014 10:32:11 +0800
Subject: [PATCH 13/22] bfs: [2/3] Refactor needs_other_cpu().

---
 kernel/sched/bfs.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index e8be229..dd71560 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -1307,9 +1307,7 @@ can_preempt(struct task_struct *p, int prio, u64 deadline)
  */
 static inline bool needs_other_cpu(struct task_struct *p, int cpu)
 {
-	if (unlikely(!cpumask_test_cpu(cpu, &p->cpus_allowed)))
-		return true;
-	return false;
+	return !cpumask_test_cpu(cpu, &p->cpus_allowed);
 }
 
 /*
-- 
2.4.6

