From:	Ashish Samant <ashish.samant@oracle.com>
To:	linux-btrfs@vger.kernel.org
Cc:	ashish.samant@oracle.com
Subject: [PATCH] btrfs: Print Warning only if ENOSPC_DEBUG is enabled
Date:	Mon, 26 Oct 2015 19:38:42 -0700

Signed-off-by : Ashish Samant <ashish.samant@oracle.com>
---
 fs/btrfs/delayed-inode.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/fs/btrfs/delayed-inode.c b/fs/btrfs/delayed-inode.c
index a2ae427..b86cfd9 100644
--- a/fs/btrfs/delayed-inode.c
+++ b/fs/btrfs/delayed-inode.c
@@ -652,9 +652,13 @@ static int btrfs_delayed_inode_reserve_metadata(
 			goto out;
 
 		ret = btrfs_block_rsv_migrate(src_rsv, dst_rsv, num_bytes);
-		if (!WARN_ON(ret))
+		if (!ret)
 			goto out;
 
+		if (btrfs_test_opt(root, ENOSPC_DEBUG))
+			WARN(1, KERN_DEBUG
+			     "btrfs: block rsv migrate returned %d\n", ret);
+
 		/*
 		 * Ok this is a problem, let's just steal from the global rsv
 		 * since this really shouldn't happen that often.
-- 
1.8.3.2
