From 7de569950716147ed436b27936628ee3ab5b45cc Mon Sep 17 00:00:00 2001
From: Con Kolivas <kernel@kolivas.org>
Date: Mon, 12 Dec 2016 17:36:44 +1100
Subject: [PATCH] Fix build for CONFIG_FREEZER disabled.x

---
 include/linux/freezer.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/linux/freezer.h b/include/linux/freezer.h
index dd03e83..2fda682 100644
--- a/include/linux/freezer.h
+++ b/include/linux/freezer.h
@@ -296,6 +296,7 @@ static inline void set_freezable(void) {}
 #define wait_event_freezekillable_unsafe(wq, condition)			\
 		wait_event_killable(wq, condition)
 
+#define pm_freezing (false)
 #endif /* !CONFIG_FREEZER */
 
 #endif	/* FREEZER_H_INCLUDED */
