From 4157360d2e1cbdfb8065f151dbe057b17188a23f Mon Sep 17 00:00:00 2001
From: Tor Vic <torvic9@mailbox.org>
Date: Mon, 7 Nov 2022 15:11:54 +0100
Subject: [PATCH] sched/alt: Add missing rq_lock_irq() function to header file

---
 kernel/sched/alt_sched.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/kernel/sched/alt_sched.h b/kernel/sched/alt_sched.h
index 93ff3bddd36f..a00bc84b93b2 100644
--- a/kernel/sched/alt_sched.h
+++ b/kernel/sched/alt_sched.h
@@ -387,6 +387,13 @@ task_rq_unlock(struct rq *rq, struct task_struct *p, struct rq_flags *rf)
 	raw_spin_unlock_irqrestore(&p->pi_lock, rf->flags);
 }
 
+static inline void
+rq_lock_irq(struct rq *rq, struct rq_flags *rf)
+	__acquires(rq->lock)
+{
+	raw_spin_lock_irq(&rq->lock);
+}
+
 static inline void
 rq_lock(struct rq *rq, struct rq_flags *rf)
 	__acquires(rq->lock)
