
Silence useless logging spam in the filesystem validator.

Signed-off-by: Holger Hoffstätte <holger@applied-asynchrony.com>

--- a/fs/fs_parser.c~	2020-03-30 00:25:41.000000000 +0200
+++ b/fs/fs_parser.c	2020-03-31 02:01:52.032965259 +0200
@@ -368,8 +368,6 @@ bool fs_validate_description(const char
 	const struct fs_parameter_spec *param, *p2;
 	bool good = true;
 
-	pr_notice("*** VALIDATE %s ***\n", name);
-
 	for (param = desc; param->name; param++) {
 		/* Check for duplicate parameter names */
 		for (p2 = desc; p2 < param; p2++) {
