From 4ecacd6ebfc92c0fcffc1da6839be0c79ce4fce1 Mon Sep 17 00:00:00 2001
From: Jens Axboe <axboe@fb.com>
Date: Wed, 30 Mar 2016 10:10:53 -0600
Subject: [PATCH 07/31] nbd: switch to using blk_queue_write_cache()

Signed-off-by: Jens Axboe <axboe@fb.com>
Reviewed-by: Christoph Hellwig <hch@lst.de>
---
 drivers/block/nbd.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/block/nbd.c b/drivers/block/nbd.c
index 08afbc7..31e73a7 100644
--- a/drivers/block/nbd.c
+++ b/drivers/block/nbd.c
@@ -693,9 +693,9 @@ static int __nbd_ioctl(struct block_device *bdev, struct nbd_device *nbd,
		if (nbd->flags & NBD_FLAG_SEND_TRIM)
			queue_flag_set_unlocked(QUEUE_FLAG_DISCARD, nbd->disk->queue);
		if (nbd->flags & NBD_FLAG_SEND_FLUSH)
-			blk_queue_flush(nbd->disk->queue, REQ_FLUSH);
+			blk_queue_write_cache(nbd->disk->queue, true, false);
		else
-			blk_queue_flush(nbd->disk->queue, 0);
+			blk_queue_write_cache(nbd->disk->queue, false, false);

		thread = kthread_run(nbd_thread_send, nbd, "%s",
				     nbd_name(nbd));
-- 
2.8.1

