
For now slightly adjust the target latency.

diff -ur linux-4.4.7.old/block/blk-wb.c linux-4.4.7/block/blk-wb.c
--- linux-4.4.7.old/block/blk-wb.c	2016-04-17 22:19:12.746140904 +0200
+++ linux-4.4.7/block/blk-wb.c	2016-04-17 22:21:09.966935326 +0200
@@ -39,7 +39,7 @@
 	 * Target min latencies, in nsecs
 	 */
 	RWB_ROT_LAT	= 75000000ULL,	/* 75 msec */
-	RWB_NONROT_LAT	= 2000000ULL,	/*   2 msec */
+	RWB_NONROT_LAT	= 3000000ULL,	/*   3 msec */
 };
 
 static inline bool rwb_enabled(struct rq_wb *rwb)
