From b006baa36c8fabdfc003161210e52d34fd6a6bc9 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 17 Nov 2017 18:21:41 +0800
Subject: [PATCH] pds: Fix rq->online is default false for cpu 0.

---
 kernel/sched/pds.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index ca0a816a9a01..88719c3737c3 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -6272,6 +6272,8 @@ void __init sched_init(void)
 		rq->iso_refractory = 0;
 		init_rq_hrtick(rq);
 	}
+	/* Set rq->online for cpu 0 */
+	cpu_rq(0)->online = true;
 
 	/*
 	 * The boot idle thread does lazy MMU switching as well:
