Silence annoying debug spam from the HID core driver.

Signed-off-by: Holger Hoffstätte <holger@applied-asynchrony.com>

--- a/drivers/hid/hid-core.c
+++ b/drivers/hid/hid-core.c
@@ -2046,8 +2046,6 @@ int hid_report_raw_event(struct hid_device *hid, enum hid_report_type type, u8 *
 		return 0;
 
 	if (unlikely(bsize < csize)) {
-		hid_warn_ratelimited(hid, "Event data for report %d is incorrect (%d vs %zu)\n",
-				     report->id, csize, bsize);
 		return -EINVAL;
 	}
 
@@ -2068,8 +2066,6 @@ int hid_report_raw_event(struct hid_device *hid, enum hid_report_type type, u8 *
 		rsize = max_buffer_size;
 
 	if (bsize < rsize) {
-		hid_warn_ratelimited(hid, "Event data for report %d was too short (%d vs %zu)\n",
-				     report->id, rsize, bsize);
 		return -EINVAL;
 	}
 
