From 23c0d91c1bd2b7a25d4030b9545d44af13744a18 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 9 Jul 2018 07:54:43 +0800
Subject: [PATCH] pds: Fix wrong nr_max_tries in migrate_pending_tasks.

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 0e909fd6988f..db16fc786065 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -3240,7 +3240,7 @@ static inline void check_deadline(struct task_struct *p, struct rq *rq)
 static inline int migrate_pending_tasks(struct rq *rq, int dest_cpu)
 {
 	int nr_migrated = 0;
-	int nr_max_tries = min(rq->nr_running, SCHED_RQ_NR_MIGRATION);
+	int nr_max_tries = min(rq->nr_running / 2, SCHED_RQ_NR_MIGRATION);
 	struct skiplist_node *node = rq->sl_header.next[0];
 
 	while (nr_max_tries && node != &rq->sl_header) {
