From bcf3c7ec5fedc1997e24602986e2c5e59e613832 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 13 Jul 2018 14:05:13 +0800
Subject: [PATCH] pds: Minor optimization in task_preemptible_rq().

---
 kernel/sched/pds.c | 17 +++++------------
 1 file changed, 5 insertions(+), 12 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 5ef6246c6f13..7508be2bc839 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -1491,22 +1491,15 @@ task_preemptible_rq(struct task_struct *p, cpumask_t *chk_mask,
 	cpumask_t tmp;
 	int level, preempt_level;
 
+#ifdef CONFIG_SCHED_SMT
+	if (cpumask_and(&tmp, chk_mask, &sched_cpu_sg_idle_mask))
+		return best_mask_cpu(task_cpu(p), &tmp);
+#endif
+
 	preempt_level = task_running_policy_level(p, this_rq());
 	level = find_first_bit(sched_rq_queued_masks_bitmap,
 			       NR_SCHED_RQ_QUEUED_LEVEL);
 
-#ifdef CONFIG_SCHED_SMT
-	if (SCHED_RQ_EMPTY == level) {
-		if (cpumask_and(&tmp, chk_mask, &sched_cpu_sg_idle_mask) ||
-		    cpumask_and(&tmp, chk_mask, &sched_rq_queued_masks[level]))
-			return best_mask_cpu(task_cpu(p), &tmp);
-
-		level = find_next_bit(sched_rq_queued_masks_bitmap,
-				      NR_SCHED_RQ_QUEUED_LEVEL,
-				      level + 1);
-	}
-#endif
-
 	while (level < preempt_level) {
 		if (cpumask_and(&tmp, chk_mask, &sched_rq_queued_masks[level]))
 			return best_mask_cpu(task_cpu(p), &tmp);
