From 2cbf07cafc7e90daed01ac183224df9848e6b457 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 15 Jul 2018 21:00:46 +0800
Subject: [PATCH] Tag PDS 0.98t

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 7508be2bc839..7cde0d7f58a3 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -94,7 +94,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.98s by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.98t by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
