From b5ce4039758eed2a8938712d157142ab774f8062 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 25 May 2018 15:49:26 +0800
Subject: [PATCH] pds: pds_sg_balance() optimization.

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 2491f85473f5..8cafded6be0e 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -2886,7 +2886,7 @@ static inline bool pds_sg_balance(struct rq *rq)
 	 * First cpu in smt group does not do smt balance, unless
 	 * other cpu is smt balance suppressed.
 	 */
-	if (cpu == cpumask_first(cpu_smt_mask(cpu)) &&
+	if (cpu == per_cpu(sd_llc_id, cpu) &&
 	    !cpumask_intersects(cpu_smt_mask(cpu), &sched_cpu_sb_suppress_mask))
 		return false;
 
