From bf28bebb62cbbdf4c333425a6ea10332242036b8 Mon Sep 17 00:00:00 2001
From: Con Kolivas <kernel@kolivas.org>
Date: Mon, 20 Feb 2017 14:00:17 +1100
Subject: [PATCH] Demote no cpumask message to info.

---
 kernel/sched/MuQSS.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/MuQSS.c b/kernel/sched/MuQSS.c
index 6f6100c..563c5a8 100644
--- a/kernel/sched/MuQSS.c
+++ b/kernel/sched/MuQSS.c
@@ -1944,7 +1944,7 @@ static int valid_task_cpu(struct task_struct *p)
 
 	if (unlikely(!cpumask_weight(&valid_mask))) {
 		/* Hotplug boot threads do this before the CPU is up */
-		WARN_ON(sched_smp_initialized);
+		printk(KERN_INFO "SCHED: No cpumask for %s/%d\n", p->comm, p->pid);
 		return cpumask_any(tsk_cpus_allowed(p));
 	}
 	return cpumask_any(&valid_mask);
